### PR TITLE
doc: update how to write plugin doc

### PR DESCRIPTION
**Pull request by stephanwlee**
This change updates our "how to write a plugin" doc. Most contents are ported from https://github.com/tensorflow/tensorboard-plugin-example.

### CODE REVIEW

**Comment by stephanwlee** [comment]
Other feedbacks are fair but I want to push back on below.

> But merging them into a more unified "walkthrough" or tutorial style might be even better.

The tensorboard-example-plugin was quite verbose and, at the end of the doc, it was hard for people to take an action other than copying the code to get started. Also I would like to note that, even with its verbose explanation on summary system, some recent contributors were not aware of well versed with the summary system; instead, people start by copying image/graph dashboards and build plugin on top of that. For that reason, I think, first, the example code should be full featured and self explanatory while keeping this doc to augment the code and explain what is a bit awkward to talk about in the code.

I am, as I am writing, coming to conclusion that this doc to appear more like a guideline on plugin authorship instead. Let me know what you think about this before I refactor the doc.

**Comment by nfelt** [comment]
I agree it shouldn't be as verbose as the old guide (although I think a significant part of why contributors didn't look at it was also that it was in a separate repo and clearly hadn't been kept up to date).  By walkthrough/tutorial I meant less something very detailed (since the code is already there for that) but more just that if we're explaining high-level concepts like frontend/backend/distribution, it's probably worth trying to ground them in the example plugin, since that's what we expect people to actually use as a starting point.

It's fine if you want to frame it as a guide to plugin authorship, that also seems like a reasonable way to approach it while keeping it relatively concise.

I do think that it's worth having at least some high-level explanation of the architecture of a plugin, since the code isn't a great way to get that, but that could also potentially be in a different doc about the plugin system as a whole (e.g. the plugin system RFC, if we fold that into a permanent part of TB documentation somewhere, which could also describe not just what plugins have to do but also the infrastructure provided by TB).

File: tensorboard/plugins/example/README.md [hunk 1]
```diff
@@ -33,16 +26,63 @@ tensorboard --logdir /tmp/whatever
...
+
+You can start building the backend by subclassing `TBPlugin` or `TBLoader` in [`base_plugin.py`]. It must have a `plugin_name` class attribute and implement the following methods:
+
+  - `is_active`: This should return whether the plugin is active (whether there exists relevant data for the plugin to process). TensorBoard will hide inactive plugins from the main navigation bar.
+  - `get_plugin_apps`: This should return a `dict` mapping route paths to WSGI applications: e.g., `"/tags"` might map to `self._serve_tags`.
+  - `define_flags`: Optional method used to expose command-line flags.
+  - `fix_flags`: : Optional method used to fix or sanitize command-line flags.
+
+[`base_plugin.py`]: https://github.com/tensorflow/tensorboard/blob/master/tensorboard/plugins/base_plugin.py
+[`EventMultiplexer`]: https://github.com/tensorflow/tensorboard/blob/master/tensorboard/backend/event_processing/event_multiplexer.py
```
[the remark this answers is not in the document]
Comment (stephanwlee) [reply tensorboard/plugins/example/README.md]: ACK about the reluctance. The change will have more profound impact on other parts of the doc that they will need to be dealt with when the time is right. For now, I will keep it as is.

File: tensorboard/plugins/example/README.md [hunk 2]
```diff
@@ -33,16 +26,63 @@ tensorboard --logdir /tmp/whatever
...
+
+A data written out as protocol buffer encodes follow: tensor, tag, step, and metadata. A Tensor is an actual value that is related to a specific tag and a specific step. A tag is a string that uniquely identifies a data series, often supplied by a user. A step encodes a temporal information which often is either batch or epoch number. Lastly, metadata can contain extra information about data and it can contain a plugin specific payload in addition to an owner identifier.
+
+### Backend: How the plugin processes data, and sends it to the browser
+
+A plugin backend is responsible for registering a plugin to TensorBoard. It also serves frontend resources and surfaces necessary data to the frontend by implementing routes (endpoints).
+
+You can start building the backend by subclassing `TBPlugin` or `TBLoader` in [`base_plugin.py`]. It must have a `plugin_name` class attribute and implement the following methods:
+
+  - `is_active`: This should return whether the plugin is active (whether there exists relevant data for the plugin to process). TensorBoard will hide inactive plugins from the main navigation bar.
```
[the remark this answers is not in the document]
Comment (stephanwlee) [reply tensorboard/plugins/example/README.md]: I, soon, want to make this non-binary and when it becomes a reality, I will update the doc to reflect the change.

**Comment by wchargin** [comment]
@googlebot: The other commits are from me; I authorize them.

**Comment by stephanwlee** [comment]
Manually used Google Doc as a linter to find any grammatical issue.

@GalOshri and @nfelt feel free to edit the README in a form of PR if you find certain parts to be insufficient or incorrect.

File: tensorboard/plugins/example/README.md [hunk 3]
```diff
@@ -19,15 +19,15 @@ TensorBoard detects plugins using the [Python `entry_points` mechanism][entrypoi
...
-  - `fix_flags`: : Optional method needed to fix or sanitize command-line flags.
+  - `fix_flags`: Optional method needed to fix or sanitize command-line flags.
 
 [entrypoints-spec]: https://packaging.python.org/specifications/entry-points/
 [entrypoints-declaration]: https://github.com/tensorflow/tensorboard/blob/373eb09e4c5d2b3cc2493f0949dc4be6b6a45e81/tensorboard/plugins/example/setup.py#L31-L35
 [`base_plugin.py`]: https://github.com/tensorflow/tensorboard/blob/master/tensorboard/plugins/base_plugin.py
 
-On instantiation, a plugin is provided a [`PluginEventMultiplexer`] object from which to read data. The `PluginRunToTagToContent` method on the multiplexer returns a dictionary containing all run–tag pairs and associated summary metadata for your plugin. For more information abuto summaries, please refer to the relevant section below.
+On instantiation, a plugin is provided a [`PluginEventMultiplexer`] object from which to read data. The `PluginRunToTagToContent` method on the multiplexer returns a dictionary containing all run–tag pairs and associated summary metadata for your plugin. For more information about summaries, please refer to the relevant section below.
```
Comment (wchargin) [inline tensorboard/plugins/example/README.md]: oops; thanks :-)

**Pull request merged**